<a href="https://colab.research.google.com/github/d-hackmt/Fundamentalss-of-Machine-Learning/blob/main/Cross_validation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### KFold Cross Validation

## Digits dataset from sklearn

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
import numpy as np
from sklearn.datasets import load_digits
import matplotlib.pyplot as plt
digits = load_digits()

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(digits.data,digits.target,test_size=0.3)

Logistic Regression

In [ ]:
lr = LogisticRegression(solver = 'liblinear',multi_class = 'ovr')
lr.fit(X_train, y_train)
lr.score(X_test, y_test)

0.9666666666666667

SVM

In [ ]:
svm = SVC(gamma = 'auto')
svm.fit(X_train, y_train)
svm.score(X_test, y_test)

0.30185185185185187

Random Forest

In [ ]:
rf = RandomForestClassifier(n_estimators=40)
rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.9740740740740741

KFold cross validation.

Basic example

In [ ]:
from sklearn.model_selection import KFold
kf = KFold(n_splits = 3)                               #no of folds on all samples of data set
kf

KFold(n_splits=3, random_state=None, shuffle=False)

In [ ]:
for train_index , test_index in kf.split([1,2,3,4,5,6,7,8,9]):                    #all will be divided in 3 different folds
  print(train_index ,test_index)

[3 4 5 6 7 8] [0 1 2]
[0 1 2 6 7 8] [3 4 5]
[0 1 2 3 4 5] [6 7 8]


Use KFold for our digits example



In [ ]:
def get_score(model , X_train , X_test ,  y_train , y_test ):                   #function to get score of our above models
  model.fit(X_train, y_train)
  return model.score(X_test, y_test)

In [ ]:
from sklearn.model_selection import StratifiedKFold
folds = StratifiedKFold(n_splits=3)

#making lists to append values

scores_logistic = []
scores_svm = []
scores_rf = []

for train_index, test_index in folds.split(digits.data,digits.target):
    X_train, X_test, y_train, y_test = digits.data[train_index], digits.data[test_index], \
                                       digits.target[train_index], digits.target[test_index]

                                      
    #writting models and appending

    scores_logistic.append(get_score(LogisticRegression(solver='liblinear',multi_class='ovr'), X_train, X_test, y_train, y_test))  
    scores_svm.append(get_score(SVC(gamma='auto'), X_train, X_test, y_train, y_test))
    scores_rf.append(get_score(RandomForestClassifier(n_estimators=40), X_train, X_test, y_train, y_test))

In [ ]:
scores_logistic

[0.8948247078464107, 0.9532554257095158, 0.9098497495826378]

In [ ]:
scores_svm

[0.3806343906510851, 0.41068447412353926, 0.5125208681135225]

In [ ]:
scores_rf

[0.9115191986644408, 0.9599332220367279, 0.9181969949916527]

SO the above was cross validation on all of our models or classifiers spliitedd and tested

cross_val_score function

In [ ]:
from sklearn.model_selection import cross_val_score

Logistic regression model performance using cross_val_score

here cv is no of folds

In [ ]:
cross_val_score(LogisticRegression(solver='liblinear',multi_class='ovr'), digits.data, digits.target,cv=3)

array([0.89482471, 0.95325543, 0.90984975])

In [ ]:
cross_val_score(SVC(gamma='auto'), digits.data, digits.target,cv=3)

array([0.38063439, 0.41068447, 0.51252087])

In [ ]:
cross_val_score(RandomForestClassifier(n_estimators=40),digits.data, digits.target,cv=3)

array([0.93155259, 0.94824708, 0.92487479])

so here all we did was use a predefined function by sklearn unlike we wrote our full fledged function !! 

Parameter tunning using k fold cross validation

In [ ]:
scores1 = cross_val_score(RandomForestClassifier(n_estimators=5),digits.data, digits.target, cv=10)
np.average(scores1)

0.8775791433891993

In [ ]:
scores2 = cross_val_score(RandomForestClassifier(n_estimators=20),digits.data, digits.target, cv=10)
np.average(scores2)

0.9471322160148976

In [ ]:
scores3 = cross_val_score(RandomForestClassifier(n_estimators=30),digits.data, digits.target, cv=10)
np.average(scores3)

0.9393513345747981

In [ ]:
scores4 = cross_val_score(RandomForestClassifier(n_estimators=40),digits.data, digits.target, cv=10)
np.average(scores4)

0.9471415270018622

Here we used cross_val_score to fine tune our random forest classifier and figured that having around 40 trees in random forest gives best result.

## Exercise

Use iris flower dataset from sklearn library and use cross_val_score against following models to measure the performance of each. In the end figure out the model with best performance,

Logistic Regression.

SVM.

Decision Tree.

Random Forest.

In [ ]:
from sklearn.datasets import load_iris
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
import numpy as np

In [ ]:
iris = load_iris()

Logistic Regression

In [ ]:
lr_scores = cross_val_score(LogisticRegression(), iris.data , iris.target)
lr_scores

/usr/local/lib/python3.8/dist-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


array([0.96666667, 1.        , 0.93333333, 0.96666667, 1.        ])

In [ ]:
np.average(lr_scores)

0.9733333333333334

Decision Tree

In [ ]:
d_scores = cross_val_score(DecisionTreeClassifier(), iris.data, iris.target)
d_scores

array([0.96666667, 0.96666667, 0.9       , 1.        , 1.        ])

In [ ]:
np.average(d_scores)

0.9666666666666668

Support Vector Machine (SVM)

In [ ]:
s_scores = cross_val_score(SVC(), iris.data, iris.target)
s_scores

array([0.96666667, 0.96666667, 0.96666667, 0.93333333, 1.        ])

In [ ]:
np.average(s_scores)

0.9666666666666666

Random Forest


In [ ]:
r_scores = cross_val_score(RandomForestClassifier(n_estimators=40), iris.data, iris.target)
r_scores

array([0.96666667, 0.96666667, 0.9       , 0.9       , 1.        ])

In [ ]:
np.average(r_scores)

0.9466666666666667

Best score so far is from Logistic Regression : 
0.9733333333333334